compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -565,7 +566,38 @@ impl<'a> Parser<'a> {
             self.bump_with((dyn_tok, dyn_tok_sp));
         }
         let ty = self.parse_ty_no_plus()?;
-        Ok(TyKind::Ref(opt_lifetime, MutTy { ty, mutbl }))
+        Ok(TyKind::Ref(opt_lifetime, pinned, MutTy { ty, mutbl }))
+    }
+
+    /// Parses `pin` and `mut` annotations on references.
+    ///
+    /// It must be either `pin const` or `pin mut`.
+    pub(crate) fn parse_pin_and_mut(&mut self) -> (Pinnedness, Mutability) {
+        let pinned = if self.eat(&TokenKind::Ident(sym::pin, IdentIsRaw::No)) {
+            Pinnedness::Pinned
+        } else {
+            Pinnedness::Not
+        };
+
+        if pinned == Pinnedness::Pinned {
+            self.psess.gated_spans.gate(sym::pin_ergonomics, self.prev_token.span);
+        }
+
+        match pinned {
+            Pinnedness::Pinned => {
+                if self.eat_keyword(kw::Const) {
+                    (pinned, Mutability::Not)
+                } else if self.eat_keyword(kw::Mut) {
+                    (pinned, Mutability::Mut)
+                } else {
+                    // FIXME: Emit an error here
```

Comment:
This should get fixed before this lands.

Reply:
Also, please in the future associate any fixmes with the feature. `FIXME(pin_ergonomics)`. It makes it easier to track.

Reply:
This is fixed now, along with tests to catch some of the more interesting subtleties here.